base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,811 @@
+                }
+
+                if (!failed)
+                {
+                    failed = !InstallFiles(from, filespec, to);
+                }
+            }
+        }
+        else
+        {
+            failed = !MoveFileEx(from, to, MOVEFILE_COPY_ALLOWED | MOVEFILE_REPLACE_EXISTING);
+            if (!failed)
+            {
+                if (Info.MainApp.IsEmpty())
+                {
+                    Info.MainApp = to;
+                }
+                failed = !AddEntry(UNOP_FILE, uninstpath);
+            }
+            else
+            {
+                ErrorBox();
+            }
+            SendMessage(g_pInfo->hDlg, IM_PROGRESS, 0, 0);
+        }
+
+        if (failed || !FindNextFileW(hFind, &wfd))
+            break;
+    }
+    FindClose(hFind);
+    return !failed;
+}
+
+static void
+AddUninstallOperationsFromDB(LPCWSTR Name, WCHAR UnOp, CStringW PathPrefix = CStringW(L""))
+{
+    CStringW item, tmp;
+    if (*GetGenerateString(Name, tmp))
+    {
+        for (int pos = 1; pos > 0;)
+        {
+            pos = tmp.Find(L'|');
+            item = pos <= 0 ? tmp : tmp.Left(pos);
+            tmp = tmp.Mid(pos + 1);
+            AddEntry(UnOp, PathPrefix + item);
+        }
+    }
+}
+
+static BOOL CALLBACK
+ExtractCallback(const EXTRACTCALLBACKINFO &, void *Cookie)
+{
+    InstallInfo &Info = *(InstallInfo *) Cookie;
+    Info.Count += 1;
+    return TRUE;
+}
+
+static DWORD CALLBACK
+ExtractAndInstallThread(LPVOID Parameter)
+{
+    const BOOL PerUserModeDefault = TRUE;
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    LPCWSTR AppName = Info.AppName, Archive = Info.ArchivePath, None = L"!";
+    CStringW installdir, tempdir, files, shortcut, tmp;
+    HRESULT hr;
+    CRegKey arpkey;
+    Info.ArpKey = &arpkey;
+    Info.Count += 1; // Cleanup
+
+    if (!*GetGenerateString(DB_GENINST_FILES, files, L"*.exe|*.*"))
+        return ErrorBox(ERROR_BAD_FORMAT);
+
+    GetCommonString(DB_SCOPE, tmp);
+    if (tmp.CompareNoCase(L"User") == 0)
+        Info.PerUser = TRUE;
+    else if (tmp.CompareNoCase(L"Machine") == 0)
+        Info.PerUser = FALSE;
+    else
+        Info.PerUser = PerUserModeDefault;
+
+    hr = GetProgramFilesPath(installdir, Info.PerUser, Info.GetGuiOwner());
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+        return ErrorBox(Info.Error);
+
+    GetGenerateString(DB_GENINST_DIR, tmp);
+    if (tmp.Find('%') == 0 && ExpandEnvStrings(tmp))
+        installdir = tmp;
+    else if (tmp.Compare(None))
+        installdir = BuildPath(installdir, tmp.IsEmpty() ? AppName : tmp.GetString());
+    Info.InstallDir = installdir.GetString();
+    Info.InstallDirLen = installdir.GetLength() + 1;
+    hr = SHPathPrepareForWriteW(Info.GetGuiOwner(), NULL, installdir, SHPPFW_DIRCREATE);
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+        return ErrorBox(Info.Error);
+
+    tempdir.Format(L"%s\\~RAM%u.tmp", installdir.GetString(), GetCurrentProcessId());
```

Comment:
Not using https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettempfilenamew ?

Reply:
GetTempFileName creates an empty dummy file I would then have to delete. This temp directory is inside the install directory that we just created, a collision is unlikely.